src/server.rs

```diff
@@ -0,0 +1,475 @@
+use std::collections::hash_map::Entry;
+use std::collections::VecDeque;
+use std::fmt;
+use std::mem::take;
+use std::sync::Arc;
+use std::task::{ready, Context, Poll};
+
+use asynchronous_codec::FramedWrite;
+use blockstore::{Blockstore, BlockstoreError};
+use cid::CidGeneric;
+use fnv::{FnvHashMap, FnvHashSet};
+use futures::future::{AbortHandle, Abortable, BoxFuture};
+use futures::stream::FuturesUnordered;
+use futures::{FutureExt, SinkExt, StreamExt};
+use libp2p_core::upgrade::ReadyUpgrade;
+use libp2p_identity::PeerId;
+use libp2p_swarm::{
+    ConnectionHandlerEvent, NotifyHandler, StreamProtocol, SubstreamProtocol, ToSwarm,
+};
+use smallvec::SmallVec;
+use tracing::{debug, info, trace, warn};
+
+use crate::incoming_stream::ServerMessage;
+use crate::message::Codec;
+use crate::proto::message::{
+    mod_Message::Block as ProtoBlock, mod_Message::Wantlist as ProtoWantlist, Message,
+};
+use crate::utils::stream_protocol;
+use crate::{Event, Result, StreamRequester, ToBehaviourEvent, ToHandlerEvent};
+
+type Sink = FramedWrite<libp2p_swarm::Stream, Codec>;
+type BlockWithCid<const S: usize> = (CidGeneric<S>, Vec<u8>);
+
+#[derive(Debug)]
+pub(crate) struct ServerBehaviour<const S: usize, B>
+where
+    B: Blockstore + Send + Send,
+{
+    protocol: StreamProtocol,
+    store: Arc<B>,
+
+    peers_waitlists: FnvHashMap<PeerId, PeerWantlist<S>>,
+    global_waitlist: FnvHashMap<CidGeneric<S>, SmallVec<[PeerId; 1]>>,
+
+    outgoing_queue: Vec<BlockWithCid<S>>,
+    outgoing_event_queue: VecDeque<ToSwarm<Event, ToHandlerEvent>>,
+
+    blockstore_tasks: FuturesUnordered<BoxFuture<'static, BlockstoreResult<S>>>,
+    blockstore_tasks_abort_handles: FnvHashMap<(PeerId, CidGeneric<S>), AbortHandle>,
+}
+
+#[derive(Debug)]
+enum BlockstoreResult<const S: usize> {
+    Get(
+        PeerId,
+        CidGeneric<S>,
+        Result<Option<Vec<u8>>, BlockstoreError>,
+    ),
+    Cancelled,
+}
+
+#[derive(Debug, Default)]
+struct PeerWantlist<const S: usize>(FnvHashSet<CidGeneric<S>>);
+
+impl<const S: usize> PeerWantlist<S> {
+    pub fn process_wantlist(&mut self, wantlist: ProtoWantlist) -> Vec<WishlistChange<S>> {
+        // XXX quietly drop invalid entries from wantlist, do we care about logging?
```

Comment:
I think we're fine to do that, either something is not a valid cid at all or it's just bigger than the behavior is configured with